File: crates/actant-core/src/hash.rs
//! Stable hashing helpers used across the substrate.

use sha2::{Digest, Sha256};

/// Canonical JSON serialization — keys sorted, no whitespace. Drives stable
/// content hashing of payloads.
pub fn canonical_json(value: &serde_json::Value) -> String {
    canonicalize(value).to_string()
}

fn canonicalize(value: &serde_json::Value) -> serde_json::Value {
    match value {
        serde_json::Value::Object(map) => {
            let mut sorted: std::collections::BTreeMap<String, serde_json::Value> =
                std::collections::BTreeMap::new();
            for (k, v) in map {
                sorted.insert(k.clone(), canonicalize(v));
            }
            serde_json::Value::Object(sorted.into_iter().collect())
        }
        serde_json::Value::Array(arr) => {
            serde_json::Value::Array(arr.iter().map(canonicalize).collect())
        }
        other => other.clone(),
    }
}

/// SHA-256 of an arbitrary string, lowercase hex.
pub fn sha256_hex(bytes: &[u8]) -> String {
    let mut h = Sha256::new();
    h.update(bytes);
    hex::encode(h.finalize())
}

/// Hash of (prev_chain_hash, payload_hash) for tamper-evident chaining.
pub fn chain_hash(prev: &str, payload_hash: &str) -> String {
    sha256_hex(format!("{prev}:{payload_hash}").as_bytes())
}

File: packages/actant-core/src/hash.ts
import { createHash } from "node:crypto";

/** Canonical JSON: stable key order, no extra whitespace. */
export function canonicalJSON(value: unknown): string {
  return JSON.stringify(sortKeys(value));
}

function sortKeys(value: unknown): unknown {
  if (Array.isArray(value)) return value.map(sortKeys);
  if (value && typeof value === "object") {
    const entries = Object.entries(value as Record<string, unknown>);
    entries.sort(([a], [b]) => (a < b ? -1 : a > b ? 1 : 0));
    return Object.fromEntries(entries.map(([k, v]) => [k, sortKeys(v)]));
  }
  return value;
}

/** SHA-256 of the canonical JSON of `value`, lowercase hex. */
export function sha256OfJSON(value: unknown): string {
  const h = createHash("sha256");
  h.update(canonicalJSON(value));
  return h.digest("hex");
}

/** SHA-256 of an arbitrary string, lowercase hex. */
export function sha256(text: string): string {
  return createHash("sha256").update(text).digest("hex");
}

/** Hash a (prev_chain_hash, payload_hash) pair into the next chain hash. */
export function nextChainHash(prevChainHash: string, payloadHash: string): string {
  return sha256(prevChainHash + ":" + payloadHash);
}
